File: articles/structural-uncertainty-and-model-form-error/fortran/structural_summary.f90
program structural_summary
  implicit none

  integer, parameter :: n = 4
  character(len=32) :: forms(n)
  real(8) :: outputs(n)
  integer :: i

  forms = (/ 'linear_decline                 ', 'proportional_decline           ', 'logistic_recovery              ', 'threshold_shift                ' /)

  do i = 1, n
    outputs(i) = simulate_model(trim(forms(i)))
  end do

  print '(A)', 'model_form projected_stock below_threshold'
  do i = 1, n
    print '(A,1X,F10.4,1X,L1)', trim(forms(i)), outputs(i), outputs(i) < 45.0d0
  end do

  print '(A,F10.4)', 'structural_spread ', maxval(outputs) - minval(outputs)

contains

  real(8) function simulate_model(form_key)
    implicit none
    character(len=*), intent(in) :: form_key
    integer :: year
    real(8) :: stock, carrying_capacity, extraction_rate, growth_rate
    real(8) :: fixed_loss, critical_threshold, growth, extraction

    stock = 80.0d0
    carrying_capacity = 120.0d0
    extraction_rate = 0.12d0
    growth_rate = 0.08d0
    fixed_loss = 5.8d0
    critical_threshold = 55.0d0

    do year = 1, 10
      if (form_key == 'linear_decline') then
        stock = max(0.0d0, stock - fixed_loss)
      else if (form_key == 'proportional_decline') then
        stock = max(0.0d0, stock - extraction_rate * stock)
      else if (form_key == 'logistic_recovery') then
        growth = growth_rate * stock * (1.0d0 - stock / carrying_capacity)
        extraction = extraction_rate * stock
        stock = max(0.0d0, stock + growth - extraction)
      else if (form_key == 'threshold_shift') then
        if (stock < critical_threshold) then
          stock = max(0.0d0, stock - 1.6d0 * extraction_rate * stock)
        else
          stock = max(0.0d0, stock - extraction_rate * stock)
        end if
      end if
    end do

    simulate_model = stock
  end function simulate_model

end program structural_summary

File: articles/structural-uncertainty-and-model-form-error/c/structural_summary.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

double simulate_model(const char *form_key) {
    double stock = 80.0;
    const double carrying_capacity = 120.0;
    const double extraction_rate = 0.12;
    const double growth_rate = 0.08;
    const double fixed_loss = 5.8;
    const double critical_threshold = 55.0;

    for (int year = 0; year < 10; ++year) {
        if (strcmp(form_key, "linear_decline") == 0) {
            stock = fmax(0.0, stock - fixed_loss);
        } else if (strcmp(form_key, "proportional_decline") == 0) {
            stock = fmax(0.0, stock - extraction_rate * stock);
        } else if (strcmp(form_key, "logistic_recovery") == 0) {
            double growth = growth_rate * stock * (1.0 - stock / carrying_capacity);
            double extraction = extraction_rate * stock;
            stock = fmax(0.0, stock + growth - extraction);
        } else if (strcmp(form_key, "threshold_shift") == 0) {
            if (stock < critical_threshold) {
                stock = fmax(0.0, stock - 1.6 * extraction_rate * stock);
            } else {
                stock = fmax(0.0, stock - extraction_rate * stock);
            }
        }
    }

    return stock;
}

int main(void) {
    const char *forms[] = {
        "linear_decline",
        "proportional_decline",
        "logistic_recovery",
        "threshold_shift"
    };

    double min_y = INFINITY;
    double max_y = -INFINITY;

    printf("model_form,projected_stock,below_threshold\n");

    for (int i = 0; i < 4; ++i) {
        double y = simulate_model(forms[i]);
        if (y < min_y) {
            min_y = y;
        }
        if (y > max_y) {
            max_y = y;
        }
        printf("%s,%.6f,%s\n", forms[i], y, y < 45.0 ? "true" : "false");
    }

    fprintf(stderr, "structural_spread=%.6f\n", max_y - min_y);
    return EXIT_SUCCESS;
}
